pluginlib/CMakeLists.txt

```diff
@@ -30,47 +30,85 @@ install(
-#         ${tinyxml2_LIBRARIES} ${catkin_LIBRARIES})
-#       set_target_properties(${PROJECT_NAME}_unique_ptr_test
-#         PROPERTIES COMPILE_FLAGS -std=c++11 LINK_FLAGS -std=c++11)
-#       add_dependencies(${PROJECT_NAME}_unique_ptr_test test_plugins)
-#     endif()
-#   endif()
-# endif()
+if(BUILD_TESTING)
+  find_package(ament_cmake_gtest REQUIRED)
+
+  set(prefix "${CMAKE_CURRENT_BINARY_DIR}/prefix")
+
+  # Write the necessary files for ament to detect a package.
+  file(COPY test/test_plugins.xml DESTINATION "${prefix}/share/test_pluginlib/")
+  configure_file(test/test_package.xml "${prefix}/share/test_pluginlib/package.xml" COPYONLY)
+
+  file(WRITE "${prefix}/share/ament_index/resource_index/test_pluginlib__pluginlib__plugin/pluginlib"
+    "share/test_pluginlib/test_plugins.xml\n")
+  file(WRITE "${prefix}/share/ament_index/resource_index/packages/test_pluginlib" "")
+
+  include_directories(include test/include)
+
+  add_library(test_plugins SHARED ./test/test_plugins.cpp)
+  target_compile_definitions(test_plugins PUBLIC "PLUGINLIB__DISABLE_BOOST_FUNCTIONS")
+  target_compile_definitions(test_plugins PRIVATE "TEST_PLUGINLIB_FIXTURE_BUILDING_LIBRARY")
+  ament_target_dependencies(
+    test_plugins
+    class_loader
+  )
+
+  # Use a generator expression so that the specified folder is used directly, without any
+  # configuration-dependent suffix.
+  #
+  # See https://cmake.org/cmake/help/v3.8/prop_tgt/LIBRARY_OUTPUT_DIRECTORY.html
+  # Ref: https://stackoverflow.com/a/45871818
+  set_target_properties(
+    test_plugins PROPERTIES
+    ARCHIVE_OUTPUT_DIRECTORY "${prefix}/lib/$<$<CONFIG:Debug>:>"
+    LIBRARY_OUTPUT_DIRECTORY "${prefix}/lib/$<$<CONFIG:Debug>:>"
+    RUNTIME_OUTPUT_DIRECTORY "${prefix}/bin/$<$<CONFIG:Debug>:>"
+  )
+
+  ament_add_gtest(${PROJECT_NAME}_unique_ptr_test
+    test/unique_ptr_test.cpp
+    APPEND_LIBRARY_DIRS "$<TARGET_FILE_DIR:test_plugins>"
+    ENV AMENT_PREFIX_PATH=${prefix}
+  )
+  if(TARGET ${PROJECT_NAME}_unique_ptr_test)
+    ament_target_dependencies(
+      ${PROJECT_NAME}_unique_ptr_test
+      class_loader
+      ament_index_cpp
+      rcutils
+      TinyXML2
+    )
+    target_link_libraries(${PROJECT_NAME}_unique_ptr_test ${TinyXML2_LIBRARIES})
+    target_compile_definitions(${PROJECT_NAME}_unique_ptr_test PUBLIC "PLUGINLIB__DISABLE_BOOST_FUNCTIONS")
+
+    if(CMAKE_COMPILER_IS_GNUCXX)
+      target_link_libraries(${PROJECT_NAME}_unique_ptr_test stdc++fs)
+    endif()
+    add_dependencies(${PROJECT_NAME}_unique_ptr_test test_plugins)
+  endif()
+
+  ament_add_gtest(${PROJECT_NAME}_utest
+    test/utest.cpp
+    APPEND_LIBRARY_DIRS "$<TARGET_FILE_DIR:test_plugins>"
+    ENV AMENT_PREFIX_PATH=${prefix}
+  )
+  if(TARGET ${PROJECT_NAME}_utest)
+    ament_target_dependencies(
+      ${PROJECT_NAME}_utest
+      class_loader
+      ament_index_cpp
+      rcutils
+      TinyXML2
+    )
+    target_link_libraries(${PROJECT_NAME}_utest ${TinyXML2_LIBRARIES})
```

Comment:
Nit, I think `ament_target_dependencies(... TinyXML2)` above does this already.

https://github.com/ament/ament_cmake/blob/2d65d41e047e203cd0531df9b0987d8aaea7d6ca/ament_cmake_target_dependencies/cmake/ament_target_dependencies.cmake#L55-L56